lib/DXIL/DxilUtil.cpp

```diff
@@ -1215,6 +1217,32 @@ Value *TryReplaceBaseCastWithGep(Value *V) {
   return nullptr;
 }
 
+// returns true if the function call is an intermediate or DXIL op 
+// that has no side effects, for functions not marked ReadNone or ReadOnly.
+bool DxilOpFunctionHasNoSideEffects(Instruction *I) {
+  if (CallInst *CI = dyn_cast<CallInst>(I)) {
+    if (CI->onlyReadsMemory()) return false;
+
+    bool isDxilOp = hlsl::OP::IsDxilOpFunc(CI->getCalledFunction());
+    if (!isDxilOp) return false;
+    switch (hlsl::OP::getOpCode(I)) {
+    case hlsl::OP::OpCode::OutputComplete: {
+      hlsl::DxilInst_OutputComplete OutputComplete(CI);
+      Value *NodeRecHandle = OutputComplete.get_output();
+      Constant *C = dyn_cast<Constant>(NodeRecHandle);
+      if (C && C->isZeroValue())
+        return true;
```

Comment:
I think Tex had in mind that this would include iconvergentmarker and the code that ensures it's a call instruction as well. I think the idea is one place to check HLSL-specific functions that aren't already marked readonly or readnone to see if they are known to not have side effects.

Reply:
Thanks for the feedback, I'll wait for Tex to confirm.

Reply:
Well, I'll go ahead and say that I think that would be a better way to structure this. Rather than having a generic named function that only checks for `OutputComplete` to determine if there are potential side effects, I think it should be more holistic and at least include the other HLSL-specific case we're aware of as well.